### PR TITLE
Working with the fetchers from both Toolforge and local pc

### PR DESCRIPTION
**Pull request by Retailed**
Hi, @tanny411 please look at the modifications to the code I made. All the necessary ports and user-password pair now can be given to the scripts to work with databases through ssh. Everything is documented with argparse and `python3 <script_name> -h` will return help info to understand it's usage better.

### CODE REVIEW

File: wikis_parser.py [hunk 1]
```diff
@@ -90,95 +93,82 @@ def get_last_update_local_db():
...
-
-
-def update_checker():
-    wiki_db_update_time = get_creation_date_from_db()
+def update_checker(meta_port=None, sources_port=None, user=None, password=None):
+    wiki_db_update_time = get_creation_date_from_db(meta_port, user, password)
     print('Wikiprojects update checker: time of last update fetched from database')
-    local_db_update_time = get_last_update_local()
-    get_last_update_local_db()
+    local_db_update_time = get_last_update_local_db(sources_port, user, password)
```
Comment (tanny411) [inline wikis_parser.py]: The name `sources_port` seems too specific to the sources table but it's port to the user_db. As in `fetch_content.py` and other files, this can be better named to indicate the database. 
Similarly are certain instances only in this file.

File: wikis_parser.py [hunk 2]
```diff
@@ -90,95 +93,82 @@ def get_last_update_local_db():
...
+    parser = argparse.ArgumentParser(
+        description="Updates wiki info stored in database in Toolforge. "
+                    "To use from local PC, use flag --local and all the additional flags needed for Ad"
+                    "establishing connection through ssh tunneling."
+                    "More help available at "
+                    "https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#SSH_tunneling_for_local_testing_which_makes_use_of_Wiki_Replica_databases"
+    )
+    parser.add_argument("--local", "-l", action="store_true",
+                        help="Connection is initiated from local pc.")
+    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc.")
```
Comment (tanny411) [inline wikis_parser.py]:
```suggestion
    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc")
```

File: db_script.py [hunk 3]
```diff
@@ -29,43 +33,67 @@ def encode_if_necessary(b):
...
+    parser = argparse.ArgumentParser(
+        description="Updates Lua scripts in database in Toolforge, fetching info from database replicas. "
+                    "To use from local PC, use flag --local and all the additional flags needed for "
+                    "establishing connection through ssh tunneling."
+                    "More help available at "
+                    "https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#SSH_tunneling_for_local_testing_which_makes_use_of_Wiki_Replica_databases"
+    )
+    parser.add_argument("--local", "-l", action="store_true",
+                        help="Connection is initiated from local pc.")
+    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc.")
```
Comment (tanny411) [inline db_script.py]:
```suggestion
    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc")
```

File: fetch_content.py [hunk 4]
```diff
@@ -285,31 +289,50 @@ def get_missed_contents(wikis):
...
+    parser.add_argument("start_idx", type=index_type,
+                        help="Starting index of info, fetched from database Sources, sorted by key (min=0).")
+    parser.add_argument("end_idx", type=index_type,
+                        help="Ending index of info, fetched from database Sources, sorted by key.(min=0)")
+    parser.add_argument("--revise", "-rev", action="store_true",
+                        help="Whether content should be revised.")
+
+    parser.add_argument("--local", "-l", action="store_true",
+                        help="Connection is initiated from local pc.")
+    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc.")
```
Comment (tanny411) [inline fetch_content.py]:
```suggestion
    local_data = parser.add_argument_group(title="Info for connecting to Toolforge from local pc")
```

**Review by tanny411 — CHANGES_REQUESTED** [review]
I believe the get_db_pages.py file is still to update with argparse?

For scripts that access multiple dbs, connecting to only metawiki should not work. Should we modify the code such that the entire code can run locally? (even though it's more appropriate to run in grid). Possibly someone can run from other cloud sources, who knows. 

It will be tough to create hundreds of db connections for each db and pass it as arguments. It may be better to connect through shards. See [toolforge:database](https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#Naming_conventions) and [this](https://wikitech.wikimedia.org/wiki/News/Wiki_Replicas_2020_Redesign#I_have_a_tool_that_queries_all_of_the_wikis_and_I_use_only_one_connection) for more on this.
In short: we can create 8 connections to access all dbs, but we have to find a way to connect to the appropriate shard for each db. 

Plus I think it would be nice to have a section in readme with commands of connecting to toolforge db. Especially how to connect to replicas, as multiple connection situation might be confusing.

ps: suggested minor changes to look better from terminal. (with -h flag)

Update: to test if connecting to metawiki works for all dbs, I ran the db_script locally and it seems to be working so far (still in progress). But I wonder how? I thought we had to connect to each cluster corresponding to that db?

File: db_script.py [hunk 5]
```diff
@@ -1,25 +1,29 @@
...
+        conn = db_acc.connect_to_user_database(DATABASE_NAME, user_db_port, user, password)
         with conn.cursor() as cur:
             for index, elem in entries.iterrows():
                 cur.execute(query,
                             [db, elem['page_id'], elem['page_title'], 1, 1])
         conn.commit()
         conn.close()
     except pymysql.err.OperationalError:
-        print('Failure: please use only in Toolforge environment')
+        print('Failure: failure, please establish connection to Toolforge')
```
Comment (tanny411) [inline db_script.py]:
```suggestion
        print('Failure: Please establish connection to Toolforge')
```

Since connection error is already raised from `connect_to_user_database`, maybe a different more specific error can be reported. Although I am not sure what, so make changes if you see fit. Similarly for other places where `connect_to_user_database` or `connect_to_replica_database`  is used.

File: db_script.py [hunk 6]
```diff
@@ -29,43 +33,67 @@ def encode_if_necessary(b):
...
+        conn = db_acc.connect_to_user_database(DATABASE_NAME, user_db_port, user, password)
         with conn.cursor() as cur:
-            cur.execute("select dbname from Sources where url is not NULL")    # all, except 'meta'
+            cur.execute("select dbname from Sources where url is not NULL")  # all, except 'meta'
             ret = [db[0] for db in cur]
         conn.close()
         return ret
     except pymysql.err.OperationalError as err:
-        print('Failure: please use only in Toolforge environment')
+        print('Failure: failure, please establish connection to Toolforge')
```
Comment (tanny411) [inline db_script.py]:
```suggestion
        print('Failure: Please establish connection to Toolforge')
```

File: utils/db_access.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                                   user=user, password=password)
+            with conn.cursor() as cur:
+                cur.execute("use " + db_name)
+            conn.commit()
+        else:
+            conn = toolforge.toolsdb(db_name)
+
+        return conn
+    except pymysql.err.OperationalError as err:
+        print('Failure: failure, please establish connection to Toolforge')
```
Comment (tanny411) [inline utils/db_access.py]:
```suggestion
        print('Failure: Please establish connection to Toolforge')
```

File: utils/db_access.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                                   user=user, password=password)
+            with conn.cursor() as cur:
+                cur.execute("use " + db_name)
+            conn.commit()
+        else:
+            conn = toolforge.connect(dbname=db_name)
+
+        return conn
+    except pymysql.err.OperationalError as err:
+        print('Failure: failure, please establish connection to Toolforge')
```
Comment (tanny411) [inline utils/db_access.py]:
```suggestion
        print('Failure: Please establish connection to Toolforge')
```

**Review by tanny411 — COMMENTED** [review]
Some minor changes.

File: wikis_parser.py (hunk 1 above)
Comment (Retailed) [reply wikis_parser.py]: > I believe the get_db_pages.py file is still to update with argparse?
> 
> For scripts that access multiple dbs, connecting to only metawiki should not work. Should we modify the code such that the entire code can run locally? (even though it's more appropriate to run in grid). Possibly someone can run from other cloud sources, who knows.
> 
> It will be tough to create hundreds of db connections for each db and pass it as arguments. It may be better to connect through shards. See [toolforge:database](https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#Naming_conventions) and [this](https://wikitech.wikimedia.org/wiki/News/Wiki_Replicas_2020_Redesign#I_have_a_tool_that_queries_all_of_the_wikis_and_I_use_only_one_connection) for more on this.
> In short: we can create 8 connections to access all dbs, but we have to find a way to connect to the appropriate shard for each db.
> 
> Plus I think it would be nice to have a section in readme with commands of connecting to toolforge db. Especially how to connect to replicas, as multiple connection situation might be confusing.
> 
> ps: suggested minor changes to look better from terminal. (with -h flag)
> 
> Update: to test if connecting to metawiki works for all dbs, I ran the db_script locally and it seems to be working so far (still in progress). But I wonder how? I thought we had to connect to each cluster corresponding to that db?

Connecting to multiple dbs was something I found working only through tests. I just noticed that when in Toolforge you connect to some database with `sql dbname`, you can choose different database by doing `use new_dbname;`. The same thing works here, I'm not really sure, how exactly, but it looks like mysql makes shard changing automatically somewhere in it's code.

Yes, it's probably makes sense to describe all the steps needed to establish the connection in README, I'm just thinking of it as of another task (it's strange to describe how to open connection to the Toolforge without mentioning that you need an account and how to create it etc., so it's kind of long)

Comment (tanny411) [reply wikis_parser.py]: Thats interesting! My script is *still* running, I just want to make sure it can connect to all dbs. But yeah, its working. Thanks!

Comment (Retailed) [reply wikis_parser.py]: > The name `sources_port` seems too specific to the sources table but it's port to the user_db. As in `fetch_content.py` and other files, this can be better named to indicate the database.
> Similarly are certain instances only in this file.

Fixed in 1536898

**Review by tanny411 — APPROVED** [review]

**Comment by tanny411** [comment]
You can merge if done.

**Comment by Retailed** [comment]
@tanny411 I've changed exception handling to more broad case in main code. Please look at 
0bbc6d5 and merge, if it's ok

**Comment by tanny411** [comment]
Looks great!
